### ISSUE TITLE
[Enhancement] Implement "Jump Back" or even "jump to previous/next position"

### ISSUE THREAD

**Issue by GitMensch** [OpenCobolIDE/OpenCobolIDE#141]
The feature "jump to definition" with [F3] / [CTRL]+[left click] is nice. What's missing is a function to jump back. Even better is the possibility to jump between different positions. The IDE would need to save the positions the cursor was in a list and two buttons/function keys would be needed to jump between the stored positions.

Possible user workflow:

* read source code
* be interested what is behind "PERFORM miracle", using [F3]
* read code like "MOVE CORRESPONDING a-struct TO b-struct"
* using [F3] to inspect definition of a-struct
* using new function to jump back
* using [F3] to inspect definition of b-struct, see that an item is missing
* using new function to jump back
* using [F3] to jump to definition of a-struct, copy the missing definition
* using new function to jump back
* using [F3] to jump to definition of b-struct, insert the missing definition
* using new function to jump back (now in miracle SECTION)
* using new function to jump back (now back in main SECTION)

I think you get the idea... I really use this a lot, jumping back 20 times back/forward some times.

**Comment by ColinDuquesnoy** [comment]
Excellent idea! I'll add that as core mode in pyqode.core, that way it will be available to other languages as well. 

We can even push it a bit further: it does not have to be linked to the goto feature, any cursor jump should be memorised (e.g. you scroll down, fix a small typo in a string and jump back to where you were).

**Comment by ColinDuquesnoy** [comment]
This will be implemented in pyqode 2.7

FYI I plan to release a new public version of OpenCobolIDE with pyqode 2.6 next weekend. I'll start working on the next pyqode version afterwards.

**Comment by GitMensch** [comment]
I meant this for every jump, just as you said.
To the additional information: I'm keen to see the new version. Do you have any interest for including a new standard cobc for windows (gnu-cobol-1.1, mingw, vbisam, pdcurses based)? If yes I try to build one.

**Comment by ColinDuquesnoy** [comment]
> To the additional information: I'm keen to see the new version. Do you have any interest for including a new standard cobc for windows (gnu-cobol-1.1, mingw, vbisam, pdcurses based)? If yes I try to build one.

Yes why not. I have not closely followed the development of GnuCOBOL, is there some major differences between OpenCOBOL 1.1 and GnuCOBOL 1.1?

Here is where I store the bundled compiler: https://github.com/OpenCobolIDE/OpenCobol-Win32-MinGW. You may want to fork it, get rid of the current files, place your new build and submit a pull request.

**Comment by GitMensch** [comment]
See http://opencobol.add1tocobol.com/gnucobol/#what-are-the-differences-between-opencobol-1-1-and-gnucobol-1-1

**Comment by ColinDuquesnoy** [comment]
Looks nice, would be nice to have it for the next release!

**Comment by ColinDuquesnoy** [comment]
> Do you have any interest for including a new standard cobc for windows (gnu-cobol-1.1, mingw, vbisam, pdcurses based)? If yes I try to build one.

Any news about this build? I am going to publish a new release this afternoon, If I can get this new build I will make it part of the next release, otherwise this will be for another time.

**Comment by GitMensch** [comment]
I've did quite an effort to get it working. After finding some bugs I've rechecked the current version (see merge request). I think I'll need half an hour to finish the work for the  GnuCOBOL 1.1 package, but likely find the time only after the children are in bed.

Am 31. Mai 2015 10:36:02 MESZ, schrieb Colin Duquesnoy <notifications@github.com>:
>> Do you have any interest for including a new standard cobc for
>windows (gnu-cobol-1.1, mingw, vbisam, pdcurses based)? If yes I try to
>build one.
>
>Any news about this build? I am going to publish a new release this
>afternoon, If I can get this new build I will make it part of the next
>release, otherwise this will be for another time.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/OpenCobolIDE/OpenCobolIDE/issues/141#issuecomment-107145845

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by ColinDuquesnoy** [comment]
I already made the release (going to move out soon) so take your time for the build ;) I can make a version 4.6.1 with your new build and the signed binaries next week end ;)

**Comment by GitMensch** [comment]
>I already made the release (going to move out soon) so take your time for the build ;) I can make a version 4.6.1 with your new build and the signed binaries next week end ;)

OK. Please try to update the online docs, too (it is labeled OCIDE 4.4.0 and there are indeed some new settings and dialog changes).

**Comment by ColinDuquesnoy** [comment]
Ok I'll do

**Comment by GitMensch** [comment]
I'd suggest to split this: one is the definition, the other two are completely different and there can be an additional jump type:

[F8] Definition

[CTRL]+[SHIFT]+[Z] Previous Location
[CTRL]+[SHIFT]+[Y] Next Location (active after using Previous Location at least once)

[CTRL]+[ALT]+[Z] Previous Changed
[CTRL]+[ALT]+[Y] Next Changed (active after using Previous Changed at least once)

**Comment by ColinDuquesnoy** [comment]
I like the idea, but instead of using an additional shortcut I will try to integrate it into the existing undo stack (this is what IntelliJ Idea/Pycharm do and this is pretty handy). I said "try" because I know there are some limitations with QPlainTextEdit and the undo stack (the stack is burried down into the private c++ implementation of QPlainTextEdit).

I just don't get what is previous/next Changed?

**Comment by GitMensch** [comment]
Previous/next Changed positions where undo/redo would do, just without the actual undo/redo part.

**Comment by ColinDuquesnoy** [comment]
> Previous/next Changed positions where undo/redo would do, just without the actual undo/redo part.

Yes for that I need to integrate the text cursor position changes into the [undo stack](http://doc.qt.io/qt-5/qundo.html)

**Comment by ColinDuquesnoy** [comment]
I have a first draft of this in pyqode 2.7.dev. Unfortunately I could not integrate the text cursor position changes into the editor's undo stack. The problem is that I don't have a direct access to the undo stack which is buried down into the private implementation of the Qt base class I use for the editor. So I had to maintain a second undo stack myself and push a new command anytime the cursor's line number changed (and you're not using jump back/jump forward).

So basically you have the following shortcuts:
- Ctrl+Alt+Z: jump back to previous cursor location
- Ctrl+Alt+Y: jump forward to the next cursor location (after one or more jump back). 

If you want to try it, add ``@dev`` at the end of each pip command in your dev script ([see](https://github.com/OpenCobolIDE/OpenCobolIDE/blob/dev/.travis.yml#L46))

The next version of both OpenCobolIDE and pyqode will be worked on in a separate dev branch (that way I can provide hotfix on the master branch). Each hotfix will be integrated into the dev branch, so if you use that branch you will have both the new hot features of v4.7 but also the bugfixes of v4.6.

**Comment by ColinDuquesnoy** [comment]
Implemented in pyqode.core (dev branch). I've been using it a lot since the last 3 months without noticing any major issues.

**Comment by GitMensch** [comment]
From your text we have
* [CTRL]+[ALT]+[Z] Previous Location
* [CTRL]+[ALT]+[Y] Next Location (active after using Previous Location at least once)

What about the second part
* [CTRL]+[SHIFT]+[Z] Previous Changed
* [CTRL]+[SHIFT]+[Y] Next Changed (active after using Previous Changed at least once)

**Comment by ColinDuquesnoy** [comment]
> What about the second part

I don't have access to the editor's undo stack data (I can only undo or redo). Therefore I could not implement that part.

**Comment by ColinDuquesnoy** [comment]
> Therefore I could not implement that part.

At least not the way I imagined. But I now have another idea that I might want to try when I have some time. I'll reopen and will let you know how that went.

**Comment by GitMensch** [comment]
Works nice but is a little buggy in 4.7-pre1 if there is no previous/next item (as it adds an uppercase Z/Y to the editor).

**Comment by ColinDuquesnoy** [comment]
I admit I haven't tried it in OCIDE only in HackEdit where this bug does not happen. I'll recheck that.

**Comment by ColinDuquesnoy** [comment]
> But I now have another idea that I might want to try when I have some time.

Tried, didn't went well. I found a hacky solution, works nice in small files but degrades performance in bigger files. Sorry I won't implement that part.

I still have to look into the bug you reported when no previous/next item.

**Comment by ColinDuquesnoy** [comment]
Bug fixed. I'll upload a new official preview build in a few minutes.

**Comment by ColinDuquesnoy** [comment]
The new build is available here on github: https://github.com/OpenCobolIDE/OpenCobolIDE/releases/tag/4.7-preview2
